// @stakeordie/griptape-vue.js/dist/wallet-store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useWalletStore = void 0;
var griptape_js_1 = require("@stakeordie/griptape.js");
var vue_1 = require("vue");
var address = vue_1.ref();
var loading = vue_1.ref(true);
var isConnected = vue_1.ref(false);
griptape_js_1.onAccountAvailable(function () {
    var theAddress = griptape_js_1.getAddress();
    isConnected.value = typeof theAddress !== 'undefined';
    address.value = theAddress;
    loading.value = false;
});
griptape_js_1.onAccountNotAvailable(function () {
    isConnected.value = false;
    loading.value = false;
});
function useWalletStore(abbrv) {
    if (abbrv === void 0) { abbrv = 12; }
    return {
        bootstrap: griptape_js_1.bootstrap,
        address: vue_1.computed(function () { return address.value; }),
        humanAddress: vue_1.computed(function () {
            if (!address.value)
                return;
            return griptape_js_1.bech32(address.value, abbrv);
        }),
        loading: vue_1.computed(function () { return loading.value; }),
        isConnected: vue_1.computed(function () { return isConnected.value; }),
    };
}
exports.useWalletStore = useWalletStore;
